include/mbedtls/gcm.h

```diff
@@ -113,21 +113,35 @@ int mbedtls_gcm_setkey( mbedtls_gcm_context *ctx,
  *                  the same as input buffer. If the buffers overlap, the output
  *                  buffer must trail at least 8 Bytes behind the input buffer.
  *
+ * \warning         When this function performs a decryption, it outputs the
+ *                  authentication tag and does not verify that the data is
+ *                  authentic. You should use this function to perform encryption
+ *                  only. For decryption, use mbedtls_gcm_auth_decrypt() instead.
+ *
  * \param ctx       The GCM context to use for encryption or decryption.
  * \param mode      The operation to perform: #MBEDTLS_GCM_ENCRYPT or
- *                  #MBEDTLS_GCM_DECRYPT.
- * \param length    The length of the input data. This must be a multiple of
- *                  16 except in the last call before mbedtls_gcm_finish().
+ *                  #MBEDTLS_GCM_DECRYPT. Note that during decryption, the
```

Comment:
This kind of contradicts previous sentence: 
>You should use this function to perform encryption only

I mean, I think we should just deprecate the decryption functionality for this functions.

Reply:
You _can_ use `mbedtls_gcm_crypt_and_tag` to decrypt, so the documentation of the function explains how the function works in that case. You _should not_ use `mbedtls_gcm_crypt_and_tag` to decrypt, so the documentation of the function recommends against it. I don't see a contradiction. How do you suggest improving the documentation?

I didn't make `mbedtls_gcm_crypt_and_tag` or `MBEDTLS_GCM_DECRYPT` officially deprecated because I wanted to get this in as fast as possible (it's confusing people), and I wanted to get the documentation into the LTS releases as well. I've filed #1674 to do the decryption after 2.10.

Reply:
@jen3andruska I would like to hear your say in this discussion

Reply:
@RonEld Yes, I think there's a contradiction here.  We should not be telling them not to use this to decrypt, and then explaining how to.  Even the above explanation @gilles-peskine-arm says: "You can use mbedtls_gcm_crypt_and_tag to decrypt...You should not use Mbedtls_gcm_crypt_and_tag to decrypt"  This is confusing for the reader, who will not understand why they are being instructed how to do something that you are advising against.  I think it needs to be one or the other.  You could say something like "We do not recommend that you use this to decrypt because...., but if you choose to, then you should be aware of...."

Reply:
@jen3andruska Thank you.
@gilles-peskine-arm Please use phrasing as suggested

Reply:
@RonEld Happy to help

Reply:
Sorry but I still don't understand the proposed change. Is that in the description of `mode`? It already recommends against using the decryption mode. Is the issue that it doesn't explicitly state the causal link from “during decryption, the tag is not verified” to “you should use this function only to encrypt”?

Reply:
I think the issue is you're casually mentioning decrypt without explicitly mentioning again that it is not recommended to use it. I'd like to suggest changing that to `#MBEDTLS_GCM_DECRYPT (not recommended, see warning above)`. This avoids repetition or fragmentation of the information leading to possible feeling of contradiction.

Also, the warning could indeed be clarified by using the structure suggested by @jen3andruska:
```
We do not recommend that you use this function to decrypt, because it does not check the authentication tag. We recommend that you use mbedtls_gcm_auth_decrypt() instead for decryption. If however you choose to use this function for decryption, you should be aware that it's your responsibility to compare the tag generated by this function to the one received with the message, in a secure way, and reject and discard the plaintext if they're not identical.
```

Reply:
I've reworded and expanded the description of the `mode` parameter. Does that correspond to what you have in mind?

Reply:
@gilles-peskine-arm this clarifies things for me. Thanks